As an APRFC user, I don't expect an IndexOutOfBoundsException while routing statistics

Given an evaluation using COWRES, submitted by APRFC
When that evaluation completes
Then I don't expect it to fail with an `IndexOutOfBoundsException`

Got the data, managed to reproduce.

```
Caused by: java.lang.IndexOutOfBoundsException: Index 30 out of bounds for length 30
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
	at java.base/java.util.Objects.checkIndex(Objects.java:361)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at wres.vis.data.ChartDataFactory.ofPairsStatistics(ChartDataFactory.java:730)
	at wres.vis.charts.ChartFactory.getPairsChart(ChartFactory.java:877)
	at wres.vis.writing.PairsStatisticsGraphicsWriter.writePlot(PairsStatisticsGraphicsWriter.java:135)
	at wres.vis.writing.PairsStatisticsGraphicsWriter.apply(PairsStatisticsGraphicsWriter.java:92)
	at wres.vis.writing.PairsStatisticsGraphicsWriter.apply(PairsStatisticsGraphicsWriter.java:37)
	at wres.datamodel.statistics.StatisticsToFormatsRouter.processPairsStatistics(StatisticsToFormatsRouter.java:890)
	at wres.datamodel.statistics.StatisticsToFormatsRouter.accept(StatisticsToFormatsRouter.java:434)
	at wres.datamodel.statistics.StatisticsToFormatsRouter.apply(StatisticsToFormatsRouter.java:318)
```

Likely something to do with an irregularly-shaped ensemble dataset, i.e., some time-series have more ensemble members than others.

--- a/src/wres/pipeline/pooling/PoolFactory.java
+++ b/src/wres/pipeline/pooling/PoolFactory.java
@@ -29,6 +29,7 @@ import java.util.stream.Stream;
 import javax.measure.Unit;
 
 import org.apache.commons.lang3.tuple.Pair;
+import org.jspecify.annotations.NonNull;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -152,7 +153,7 @@ public class PoolFactory
     private final AtomicLong poolId = new AtomicLong( 0 );
 
     /** Cache of unit mappers for re-use. **/
-    private final Cache<String, DoubleUnaryOperator> converterCache =
+    private final Cache<@NonNull String, DoubleUnaryOperator> converterCache =
             Caffeine.newBuilder()
                     .maximumSize( 10 )
                     .build();
@@ -741,43 +742,38 @@ public class PoolFactory
 
         // Build and return the pool suppliers
         List<Supplier<Pool<TimeSeries<Pair<Double, Double>>>>> rawSuppliers =
-                new PoolsGenerator.Builder<Double, Double, Double>().setProject( project )
-                                                                    .setRetrieverFactory( retrieverFactory )
-                                                                    .setPoolRequests( poolRequests )
-                                                                    .setBaselineGenerator( baselineGenerator )
-                                                                    .setLeftTransformerPreRescaling(
-                                                                            preRescalingTransformer )
-                                                                    .setRightTransformerPreRescaling(
-                                                                            preRescalingTransformer )
-                                                                    .setBaselineTransformerPreRescaling(
-                                                                            preRescalingTransformer )
-                                                                    .setLeftTransformerPostRescaling(
-                                                                            composedLeftTransformer )
-                                                                    .setRightTransformerPostRescaling(
-                                                                            composedRightTransformer )
-                                                                    .setBaselineTransformerPostRescaling(
-                                                                            composedBaselineTransformer )
-                                                                    .setLeftFilter( filter )
-                                                                    .setRightFilter( filter )
-                                                                    .setBaselineFilter( filter )
-                                                                    .setLeftMissingFilter( missingFilter )
-                                                                    .setRightMissingFilter( missingFilter )
-                                                                    .setBaselineMissingFilter( missingFilter )
-                                                                    .setLeftUpscaler( leftUpscaler )
-                                                                    .setRightUpscaler( rightUpscaler )
-                                                                    .setCovariateFilters( covariateFilters )
-                                                                    .setBaselineUpscaler( baselineUpscaler )
-                                                                    .setLeftTimeShift( leftTimeShift )
-                                                                    .setRightTimeShift( rightTimeShift )
-                                                                    .setBaselineTimeShift( baselineTimeShift )
-                                                                    .setPairer( pairer )
-                                                                    .setPairFrequency( pairFrequency )
-                                                                    .setCrossPairer( crossPairer )
-                                                                    .setClimateMapper( Double::doubleValue )
-                                                                    .setClimateAdmissibleValue( Double::isFinite )
-                                                                    .setBaselineShim( Function.identity() )
-                                                                    .build()
-                                                                    .get();
+                new PoolsGenerator.Builder<Double, Double, Double>()
+                        .setProject( project )
+                        .setRetrieverFactory( retrieverFactory )
+                        .setPoolRequests( poolRequests )
+                        .setBaselineGenerator( baselineGenerator )
+                        .setLeftTransformerPreRescaling( preRescalingTransformer )
+                        .setRightTransformerPreRescaling( preRescalingTransformer )
+                        .setBaselineTransformerPreRescaling( preRescalingTransformer )
+                        .setLeftTransformerPostRescaling( composedLeftTransformer )
+                        .setRightTransformerPostRescaling( composedRightTransformer )
+                        .setBaselineTransformerPostRescaling( composedBaselineTransformer )
+                        .setLeftFilter( filter )
+                        .setRightFilter( filter )
+                        .setBaselineFilter( filter )
+                        .setLeftMissingFilter( missingFilter )
+                        .setRightMissingFilter( missingFilter )
+                        .setBaselineMissingFilter( missingFilter )
+                        .setLeftUpscaler( leftUpscaler )
+                        .setRightUpscaler( rightUpscaler )
+                        .setCovariateFilters( covariateFilters )
+                        .setBaselineUpscaler( baselineUpscaler )
+                        .setLeftTimeShift( leftTimeShift )
+                        .setRightTimeShift( rightTimeShift )
+                        .setBaselineTimeShift( baselineTimeShift )
+                        .setPairer( pairer )
+                        .setPairFrequency( pairFrequency )
+                        .setCrossPairer( crossPairer )
+                        .setClimateMapper( Double::doubleValue )
+                        .setClimateAdmissibleValue( Double::isFinite )
+                        .setBaselineShim( Function.identity() )
+                        .build()
+                        .get();
 
         return this.getComposedSuppliers( poolRequests, rawSuppliers );
     }
@@ -880,11 +876,12 @@ public class PoolFactory
         Function<Geometry, DoubleUnaryOperator> rightOffsetGenerator =
                 this.getOffsetTransformer( rightOffsets, DatasetOrientation.RIGHT );
         Dataset right = DeclarationUtilities.getDeclaredDataset( project.getDeclaration(), DatasetOrientation.RIGHT );
-        UnaryOperator<Event<Ensemble>> rightValueTransformer = this.getValueTransformer( leftValueTransformer,
-                                                                                         right,
-                                                                                         cacheSorted );
+        UnaryOperator<Event<Ensemble>> rightValueTransformer = this.getEnsembleValueTransformer( leftValueTransformer,
+                                                                                                 right,
+                                                                                                 cacheSorted );
+
         UnaryOperator<TimeSeries<Ensemble>> rightValueAndUnitTransformer =
-                this.getValueTransformer( rightOffsetGenerator, rightValueTransformer );
+                this.getEnsembleValueTransformer( rightOffsetGenerator, rightValueTransformer );
 
         UnaryOperator<TimeSeries<Ensemble>> rightValidTimeSeasonTransformer =
                 this.getValidTimeSeasonTransformer( declaration.season(),
@@ -910,12 +907,13 @@ public class PoolFactory
                                                                 DatasetOrientation.BASELINE );
         }
 
-        UnaryOperator<Event<Ensemble>> baselineValueTransformer = this.getValueTransformer( leftValueTransformer,
-                                                                                            baseline,
-                                                                                            cacheSorted );
+        UnaryOperator<Event<Ensemble>> baselineValueTransformer =
+                this.getEnsembleValueTransformer( leftValueTransformer,
+                                                  baseline,
+                                                  cacheSorted );
 
         UnaryOperator<TimeSeries<Ensemble>> baselineValueAndUnitTransformer =
-                this.getValueTransformer( baselineOffsetGenerator, baselineValueTransformer );
+                this.getEnsembleValueTransformer( baselineOffsetGenerator, baselineValueTransformer );
 
         // Re-use the right season transformer
         UnaryOperator<TimeSeries<Ensemble>> composedBaselineTransformer =
@@ -1057,11 +1055,11 @@ public class PoolFactory
                 this.getOffsetTransformer( rightOffsets, DatasetOrientation.RIGHT );
 
         Dataset right = DeclarationUtilities.getDeclaredDataset( project.getDeclaration(), DatasetOrientation.RIGHT );
-        UnaryOperator<Event<Ensemble>> rightValueTransformer = this.getValueTransformer( leftValueTransformer,
-                                                                                         right,
-                                                                                         cacheSorted );
+        UnaryOperator<Event<Ensemble>> rightValueTransformer = this.getEnsembleValueTransformer( leftValueTransformer,
+                                                                                                 right,
+                                                                                                 cacheSorted );
         UnaryOperator<TimeSeries<Ensemble>> rightValueAndUnitTransformer =
-                this.getValueTransformer( rightOffsetGenerator, rightValueTransformer );
+                this.getEnsembleValueTransformer( rightOffsetGenerator, rightValueTransformer );
 
         // Baseline transformer, which is a composition of several transformations
         Dataset baseline = null;
@@ -1078,12 +1076,13 @@ public class PoolFactory
         Function<Geometry, DoubleUnaryOperator> baselineOffsetGenerator =
                 this.getOffsetTransformer( baselineOffsets, DatasetOrientation.BASELINE );
 
-        UnaryOperator<Event<Ensemble>> baselineValueTransformer = this.getValueTransformer( leftValueTransformer,
-                                                                                            baseline,
-                                                                                            cacheSorted );
+        UnaryOperator<Event<Ensemble>> baselineValueTransformer =
+                this.getEnsembleValueTransformer( leftValueTransformer,
+                                                  baseline,
+                                                  cacheSorted );
 
         UnaryOperator<TimeSeries<Ensemble>> baselineValueAndUnitTransformer =
-                this.getValueTransformer( baselineOffsetGenerator, baselineValueTransformer );
+                this.getEnsembleValueTransformer( baselineOffsetGenerator, baselineValueTransformer );
 
         // Currently only a seasonal filter, which applies equally to all sides
         Predicate<TimeSeries<Double>> singleValuedFilter = this.getReferenceTimeSeasonFilter( declaration.season() );
@@ -1434,8 +1433,8 @@ public class PoolFactory
      * @return a transformer that applies a unit conversion followed by the input transformer
      */
 
-    private UnaryOperator<TimeSeries<Ensemble>> getValueTransformer( Function<Geometry, DoubleUnaryOperator> offsetGenerator,
-                                                                     UnaryOperator<Event<Ensemble>> basicTransformer )
+    private UnaryOperator<TimeSeries<Ensemble>> getEnsembleValueTransformer( Function<Geometry, DoubleUnaryOperator> offsetGenerator,
+                                                                             UnaryOperator<Event<Ensemble>> basicTransformer )
     {
         return toTransform -> {
             // Apply the unit mapping first, then the basic transformer
@@ -1660,9 +1659,9 @@ public class PoolFactory
      * @return a transformer or null
      */
 
-    private UnaryOperator<Event<Ensemble>> getValueTransformer( DoubleUnaryOperator valueTransformer,
-                                                                Dataset dataset,
-                                                                boolean cacheSorted )
+    private UnaryOperator<Event<Ensemble>> getEnsembleValueTransformer( DoubleUnaryOperator valueTransformer,
+                                                                        Dataset dataset,
+                                                                        boolean cacheSorted )
     {
         List<String> inclusive = null;
         List<String> exclusive = null;
@@ -1694,8 +1693,10 @@ public class PoolFactory
             Labels ensLabels = returnMe.getLabels();
 
             // If filters were defined, they must select something
-            if ( ( ( Objects.nonNull( finalInclusive ) && !finalInclusive.isEmpty() )
-                   || ( Objects.nonNull( finalExclusive ) && !finalExclusive.isEmpty() ) )
+            if ( ( ( Objects.nonNull( finalInclusive )
+                     && !finalInclusive.isEmpty() )
+                   || ( Objects.nonNull( finalExclusive )
+                        && !finalExclusive.isEmpty() ) )
                  && !returnMe.hasLabels() )
             {
                 throw new IllegalArgumentException( "When attempting to filter ensemble members, discovered some "
@@ -1720,9 +1721,8 @@ public class PoolFactory
                                                    finalInclusive,
                                                    finalExclusive );
 
-            // Remove any missing ensemble members
-            returnMe = Slicer.eachOfRight( MissingValues::isNotMissingValue )
-                             .apply( returnMe );
+            // Add default ensemble labels, if needed
+            returnMe = this.addEnsembleLabelsIfMissing( returnMe );
 
             // Cache the sorted ensemble members if required
             if ( cacheSorted )
@@ -1732,6 +1732,24 @@ public class PoolFactory
 
             return Event.of( toTransform.getTime(), returnMe );
         };
+    }
+
+    /**
+     * Adds default ensemble labels if missing.
+     *
+     * @param ensemble the ensemble
+     * @return the ensemble with default labels
+     */
+
+    private Ensemble addEnsembleLabelsIfMissing( Ensemble ensemble )
+    {
+        if ( !ensemble.hasLabels() )
+        {
+            ensemble = Ensemble.of( ensemble.getMembers(),
+                                    true );
+        }
+
+        return ensemble;
     }
 
     /**
@@ -1771,7 +1789,9 @@ public class PoolFactory
             }
 
             // If inclusive filters were defined, they must select something
-            if ( Objects.nonNull( finalInclusive ) && !finalInclusive.isEmpty() && valuesToUse.isEmpty() )
+            if ( Objects.nonNull( finalInclusive )
+                 && !finalInclusive.isEmpty()
+                 && valuesToUse.isEmpty() )
             {
                 throw new IllegalArgumentException( "When attempting to filter ensemble members, discovered some "
                                                     + "inclusive filters, but no members matches these filters, "

--- a/src/wres/pipeline/statistics/EnsembleStatisticsProcessor.java
+++ b/src/wres/pipeline/statistics/EnsembleStatisticsProcessor.java
@@ -11,6 +11,7 @@ import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Future;
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.function.Supplier;
 import java.util.function.ToDoubleFunction;
 import java.util.function.UnaryOperator;
 import java.util.stream.Collectors;
@@ -293,27 +294,41 @@ public class EnsembleStatisticsProcessor extends StatisticsProcessor<Pool<TimeSe
         Pool<Pair<Double, Ensemble>> unpacked =
                 this.doWorkWithSlicingExecutor( () -> PoolSlicer.unpack( adjustedPool ) );
 
+        // Remove any missing ensemble members from the unpacked pool only, preserving the original structure for the
+        // time-series metrics. See GitHub #711. This step was previously abstracted to the PoolFactory for performance
+        // reasons when conducting sampling uncertainty estimation. See Redmine #118831-88. However, it is convenient
+        // to preserve the original time-series structure for time-series metrics. This structure cannot be
+        // reconstructed across pools when each pool is processed separately, which may lead to inconsistencies
+        // (e.g., in the ensemble series/labels present across pools). Thus, the filtering was restored here to address
+        // GitHub #711, allowing the unpacked pool to have missing values removed and the timed pool to retain its
+        // original structure.
+        Supplier<Pool<Pair<Double, Ensemble>>> slicingTask =
+                () -> PoolSlicer.transform( unpacked,
+                                            Slicer.leftAndEachOfRight( MissingValues::isNotMissingValue ),
+                                            metaMapper );
+        Pool<Pair<Double, Ensemble>> unpackedNoMissing = this.doWorkWithSlicingExecutor( slicingTask );
 
         LOGGER.debug( "Computing ensemble statistics from {} pairs.",
-                      unpacked.get()
-                              .size() );
+                      unpackedNoMissing.get()
+                                       .size() );
 
         // Process the metrics that consume ensemble pairs
         if ( this.hasMetrics( SampleDataGroup.ENSEMBLE ) )
         {
-            this.processEnsemblePairs( unpacked, futures );
+            this.processEnsemblePairs( unpackedNoMissing, futures );
         }
 
         // Process the metrics that consume time-series of ensemble pairs
         if ( this.hasMetrics( SampleDataGroup.ENSEMBLE_TIME_SERIES ) )
         {
+            // Note that missing values are still in place for this dataset. See above and GitHub #711.
             this.processEnsembleTimeSeriesPairs( adjustedPool, futures );
         }
 
         // Process the metrics that consume discrete probability pairs derived from the ensemble pairs
         if ( this.hasMetrics( SampleDataGroup.DISCRETE_PROBABILITY ) )
         {
-            this.processDiscreteProbabilityPairs( unpacked, futures );
+            this.processDiscreteProbabilityPairs( unpackedNoMissing, futures );
         }
 
         // Process the metrics that consume dichotomous pairs derived from the ensemble pairs
@@ -323,7 +338,7 @@ public class EnsembleStatisticsProcessor extends StatisticsProcessor<Pool<TimeSe
             LOGGER.debug( "Encountered dichotomous metrics and decision thresholds, which means that dichtomous "
                           + "metrics will be computed for the ensemble pairs." );
 
-            this.processPairsForDichotomousMetrics( unpacked, futures );
+            this.processPairsForDichotomousMetrics( unpackedNoMissing, futures );
         }
 
         // Process the ensemble result, which do not yet include single-valued metrics
@@ -850,7 +865,8 @@ public class EnsembleStatisticsProcessor extends StatisticsProcessor<Pool<TimeSe
                                                                     StatisticsStore.Builder futures )
     {
         // Don't waste cpu cycles computing statistics for empty pairs
-        if ( pairs.get().isEmpty() )
+        if ( pairs.get()
+                  .isEmpty() )
         {
             LOGGER.debug( SKIPPING_THE_CALCULATION_OF_STATISTICS_FOR_AN_EMPTY_POOL_OF_PAIRS_WITH_METADATA,
                           pairs.getMetadata() );

--- a/src/wres/pipeline/statistics/SingleValuedStatisticsProcessor.java
+++ b/src/wres/pipeline/statistics/SingleValuedStatisticsProcessor.java
@@ -266,8 +266,10 @@ public class SingleValuedStatisticsProcessor extends StatisticsProcessor<Pool<Ti
 
         // Log
         LOGGER.debug( PROCESSING_COMPLETE_MESSAGE,
-                      pool.getMetadata().getFeatureGroup(),
-                      pool.getMetadata().getTimeWindow() );
+                      pool.getMetadata()
+                          .getFeatureGroup(),
+                      pool.getMetadata()
+                          .getTimeWindow() );
 
         // Process and return the result
         return futures.build();

--- a/test/wres/pipeline/statistics/EnsembleStatisticsProcessorTest.java
+++ b/test/wres/pipeline/statistics/EnsembleStatisticsProcessorTest.java
@@ -1951,8 +1951,8 @@ class EnsembleStatisticsProcessorTest
         Pairs pairs = Pairs.newBuilder()
                            .addLeftVariableNames( DatasetOrientation.LEFT.toString()
                                                                          .toUpperCase() )
-                           .addRightVariableNames( "MEMBER 1" )
-                           .addRightVariableNames( "MEMBER 2" )
+                           .addRightVariableNames( "1" )
+                           .addRightVariableNames( "2" )
                            .addTimeSeries( timeSeries )
                            .build();
 

--- a/test/wres/pipeline/statistics/TestDataFactory.java
+++ b/test/wres/pipeline/statistics/TestDataFactory.java
@@ -807,7 +807,9 @@ public final class TestDataFactory
     public static wres.datamodel.pools.Pool<TimeSeries<Pair<Double, Ensemble>>> getTimeSeriesOfEnsemblePairsThree()
     {
         SortedSet<Event<Pair<Double, Ensemble>>> values = new TreeSet<>();
-        values.add( Event.of( Instant.parse( "1985-03-12T00:00:00Z" ), Pair.of( 22.9, Ensemble.of( 22.8, 23.9 ) ) ) );
+        values.add( Event.of( Instant.parse( "1985-03-12T00:00:00Z" ),
+                              Pair.of( 22.9, Ensemble.of( new double[] { 22.8, 23.9 },
+                                                          Ensemble.Labels.of( "1", "2" ) ) ) ) );
 
         TimeWindow inner = MessageUtilities.getTimeWindow( Instant.parse( FIRST_TIME ),
                                                            Instant.parse( SECOND_TIME ),

--- a/wres-datamodel/src/wres/datamodel/Slicer.java
+++ b/wres-datamodel/src/wres/datamodel/Slicer.java
@@ -8,10 +8,12 @@ import java.util.Collection;
 import java.util.Collections;
 import java.util.Comparator;
 import java.util.EnumMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.OptionalDouble;
+import java.util.Set;
 import java.util.SortedSet;
 import java.util.TreeSet;
 import java.util.function.BiFunction;
@@ -491,6 +493,41 @@ public final class Slicer
         return input.stream()
                     .collect( Collectors.groupingBy( pair -> pair.getRight()
                                                                  .size() ) );
+    }
+
+    /**
+     * Enforces a common set of ensemble member labels, padding with missing values where needed.
+     * @param unpadded the unpadded ensemble
+     * @param labels the labels to enforce
+     * @return the padded series
+     */
+
+    public static Ensemble pad( Ensemble unpadded, SortedSet<String> labels )
+    {
+        Objects.requireNonNull( unpadded );
+        Objects.requireNonNull( labels );
+
+        Ensemble.Labels nextLabels = unpadded.getLabels();
+        Set<String> current = new HashSet<>( Arrays.asList( nextLabels.getLabels() ) );
+        Set<String> difference = new HashSet<>( labels );
+        difference.removeAll( current );
+
+        int index = 0;
+        double[] members = new double[labels.size()];
+        for ( String nextLabel : labels )
+        {
+            if ( difference.contains( nextLabel ) )
+            {
+                members[index] = MissingValues.DOUBLE;
+            }
+            else
+            {
+                members[index] = unpadded.getMember( nextLabel );
+            }
+            index++;
+        }
+        Ensemble.Labels newLabels = Ensemble.Labels.of( labels.toArray( new String[0] ) );
+        return Ensemble.of( members, newLabels );
     }
 
     /**

--- a/wres-datamodel/src/wres/datamodel/statistics/PairsStatisticOuter.java
+++ b/wres-datamodel/src/wres/datamodel/statistics/PairsStatisticOuter.java
@@ -84,6 +84,12 @@ public class PairsStatisticOuter implements Statistic<PairsStatistic>
                                       .getName() );
 
         builder.append( "metadata", this.getPoolMetadata() );
+        builder.append( "leftNames", this.getStatistic()
+                                         .getStatistics()
+                                         .getLeftVariableNamesList() );
+        builder.append( "rightNames", this.getStatistic()
+                                         .getStatistics()
+                                         .getRightVariableNamesList() );
 
         return builder.toString();
     }

--- a/wres-datamodel/src/wres/datamodel/time/TimeSeriesSlicer.java
+++ b/wres-datamodel/src/wres/datamodel/time/TimeSeriesSlicer.java
@@ -922,10 +922,10 @@ public final class TimeSeriesSlicer
      *
      * <p>Reference datetimes metadata are lost but ensemble labels are preserved.
      *
-     * @param timeSeries The time-series to decompose
-     * @return A map with the trace label as key, sorted set of Events as value.
-     * @throws UnsupportedOperationException When the ensemble events contain a varying number of members and the 
-     *            ensemble labels have not been provided to distinguish between them.
+     * @param timeSeries the time-series to decompose
+     * @return a map with the trace label as key, sorted set of Events as value
+     * @throws UnsupportedOperationException when the ensemble events contain a varying number of members and the
+     *            ensemble labels have not been provided to distinguish between them
      */
 
     public static Map<Object, SortedSet<Event<Double>>> decomposeWithLabels( TimeSeries<Ensemble> timeSeries )
@@ -946,7 +946,8 @@ public final class TimeSeriesSlicer
         for ( Event<Ensemble> next : timeSeries.getEvents() )
         {
             // No labels, so check for a constant number of ensemble members
-            if ( Objects.nonNull( traceCount ) && next.getValue().size() != traceCount )
+            if ( Objects.nonNull( traceCount ) && next.getValue()
+                                                      .size() != traceCount )
             {
                 throw new UnsupportedOperationException( "Cannot determine the ensemble traces from the input "
                                                          + "time-series because the number of ensemble members "

--- a/wres-datamodel/src/wres/datamodel/types/Ensemble.java
+++ b/wres-datamodel/src/wres/datamodel/types/Ensemble.java
@@ -8,6 +8,7 @@ import java.util.concurrent.atomic.AtomicReference;
 
 import com.github.benmanes.caffeine.cache.Cache;
 import com.github.benmanes.caffeine.cache.Caffeine;
+import org.jspecify.annotations.NonNull;
 
 /**
  * Stores an array of ensemble member values as double values and, optionally, an array of ensemble member labels.
@@ -21,9 +22,9 @@ public class Ensemble implements Comparable<Ensemble>
     {
         /** A cache of ensemble names to re-use, indexed by the concatenation of the names as a unique identifier.
          * Allow one hundred; there should not be more than a handful of instances across recent evaluations. */
-        private static final Cache<String, Labels> LABELS_CACHE = Caffeine.newBuilder()
-                                                                          .maximumSize( 100 )
-                                                                          .build();
+        private static final Cache<@NonNull String, Labels> LABELS_CACHE = Caffeine.newBuilder()
+                                                                                   .maximumSize( 100 )
+                                                                                   .build();
 
         /** Empty labels. */
         private static final Labels EMPTY_LABELS = new Labels( new String[0] );
@@ -185,6 +186,36 @@ public class Ensemble implements Comparable<Ensemble>
                                Labels labels )
     {
         return new Ensemble( members, labels, false );
+    }
+
+    /**
+     * Returns a {@link Ensemble} from a primitive array of members and creates default labels for the ensemble members
+     * on request.
+     *
+     * @param members the ensemble members
+     * @param defaultLabels is true to add default labels.
+     * @return the ensemble
+     * @throws NullPointerException if the members are null
+     */
+
+    public static Ensemble of( final double[] members,
+                               boolean defaultLabels )
+    {
+        if ( !defaultLabels )
+        {
+            return Ensemble.of( members, Labels.EMPTY_LABELS, false );
+        }
+
+        Objects.requireNonNull( members );
+
+        int count = members.length;
+        String[] labels = new String[count];
+        for ( int i = 0; i < count; i++ )
+        {
+            labels[i] = "MEMBER " + ( i + 1 );
+        }
+
+        return new Ensemble( members, Ensemble.Labels.of( labels ), false );
     }
 
     /**
@@ -412,7 +443,7 @@ public class Ensemble implements Comparable<Ensemble>
     {
         Objects.requireNonNull( members );
 
-        if( Objects.isNull( labels ) )
+        if ( Objects.isNull( labels ) )
         {
             labels = Labels.EMPTY_LABELS;
         }

--- a/wres-datamodel/test/wres/datamodel/SlicerTest.java
+++ b/wres-datamodel/test/wres/datamodel/SlicerTest.java
@@ -8,6 +8,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.SortedSet;
 import java.util.TreeSet;
 import java.util.function.DoublePredicate;
 import java.util.function.DoubleUnaryOperator;
@@ -874,6 +875,25 @@ class SlicerTest
                    () -> assertEquals( 12.0, transformerTwo.applyAsDouble( 16.0 ), 1 ),
                    () -> assertEquals( 12.0, transformerThree.applyAsDouble( 16.0 ), 1 ),
                    () -> assertEquals( 99.0, transformerThree.applyAsDouble( 24.0 ), 1 ) );
+    }
+
+    @Test
+    void testPad()
+    {
+        Labels labels = Labels.of( "2", "3" );
+        Ensemble unpadded = Ensemble.of( new double[] { 2, 3 },
+                                         labels );
+        SortedSet<String> expandedLabels = new TreeSet<>();
+        expandedLabels.add( "1" );
+        expandedLabels.add( "2" );
+        expandedLabels.add( "3" );
+        expandedLabels.add( "4" );
+
+        Ensemble actual = Slicer.pad( unpadded, expandedLabels );
+        Labels expectedLabels = Labels.of( "1", "2", "3", "4" );
+        Ensemble expected = Ensemble.of( new double[] { MissingValues.DOUBLE, 2, 3, MissingValues.DOUBLE },
+                                         expectedLabels );
+        assertEquals( expected, actual );
     }
 
 }

--- a/wres-datamodel/test/wres/datamodel/types/EnsembleTest.java
+++ b/wres-datamodel/test/wres/datamodel/types/EnsembleTest.java
@@ -304,6 +304,15 @@ final class EnsembleTest
     }
 
     @Test
+    void testDefaultLabels()
+    {
+        Ensemble actual = Ensemble.of( new double[] { 1, 2, 3, 4 }, true );
+        Ensemble expected = Ensemble.of( new double[] { 1, 2, 3, 4 },
+                                         Labels.of( "MEMBER 1", "MEMBER 2", "MEMBER 3", "MEMBER 4" ) );
+        assertEquals( expected, actual );
+    }
+
+    @Test
     void checkForExpectedExceptionOnConstructionWhenThereAreMoreLabelsThanMembers()
     {
         Labels labels = Labels.of( "A", "B" );

--- a/wres-metrics/src/wres/metrics/timeseries/SpaghettiPlot.java
+++ b/wres-metrics/src/wres/metrics/timeseries/SpaghettiPlot.java
@@ -1,18 +1,20 @@
 package wres.metrics.timeseries;
 
 import java.time.Instant;
-import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.SortedSet;
+import java.util.TreeSet;
 import java.util.function.Function;
+import java.util.stream.Collectors;
 
 import org.apache.commons.lang3.tuple.Pair;
 
 import wres.config.MetricConstants;
 import wres.config.components.DatasetOrientation;
+import wres.datamodel.Slicer;
 import wres.datamodel.messages.MessageFactory;
 import wres.datamodel.pools.Pool;
 import wres.datamodel.pools.PoolException;
@@ -21,6 +23,7 @@ import wres.datamodel.time.Event;
 import wres.datamodel.time.TimeSeries;
 import wres.datamodel.types.Ensemble;
 import wres.metrics.Metric;
+import wres.metrics.MetricCalculationException;
 import wres.statistics.MessageUtilities;
 import wres.statistics.generated.MetricName;
 import wres.statistics.generated.Pairs;
@@ -39,9 +42,6 @@ public class SpaghettiPlot implements Metric<Pool<TimeSeries<Pair<Double, Ensemb
     private static final PairsMetric PAIRS_METRIC = PairsMetric.newBuilder()
                                                                .setName( MetricName.SPAGHETTI_PLOT )
                                                                .build();
-
-    /** Start to default label string. */
-    private static final String MEMBER = "MEMBER ";
 
     /**
      * Returns an instance.
@@ -70,19 +70,32 @@ public class SpaghettiPlot implements Metric<Pool<TimeSeries<Pair<Double, Ensemb
                                                                               .isEmpty() )
                                                               .toList();
 
+        this.validateLabels( series );
+
+        // Find the superset of labels across all series
+        SortedSet<String> stringLabels = series.stream()
+                                               .flatMap( t -> t.getEvents()
+                                                               .stream()
+                                                               .flatMap( e -> Arrays.stream( e.getValue()
+                                                                                              .getRight()
+                                                                                              .getLabels()
+                                                                                              .getLabels() ) ) )
+                                               .collect( Collectors.toCollection( TreeSet::new ) );
+
         // Add the variable names
         if ( !series.isEmpty() )
         {
             pairs.addLeftVariableNames( DatasetOrientation.LEFT.toString()
                                                                .toUpperCase() );
-            List<String> labels = this.getEnsembleLabels( series );
-            pairs.addAllRightVariableNames( labels );
+            pairs.addAllRightVariableNames( stringLabels );
         }
 
         for ( TimeSeries<Pair<Double, Ensemble>> nextSeries : series )
         {
+
             Pairs.TimeSeriesOfPairs.Builder nextBuilder = Pairs.TimeSeriesOfPairs.newBuilder();
             Map<ReferenceTime.ReferenceTimeType, Instant> referenceTimes = nextSeries.getReferenceTimes();
+
             // Add the reference times
             referenceTimes.entrySet()
                           .stream()
@@ -97,9 +110,13 @@ public class SpaghettiPlot implements Metric<Pool<TimeSeries<Pair<Double, Ensemb
                                                                      .getLeft() )
                                                        .setValidTime( MessageUtilities.getTimestamp( pair.getTime() ) );
 
-                for ( double nextMember : pair.getValue()
-                                              .getRight()
-                                              .getMembers() )
+                Ensemble unpadded = pair.getValue()
+                                        .getRight();
+
+                // Pad the ensemble with a consistent set of members. See GitHub #711.
+                Ensemble padded = Slicer.pad( unpadded, stringLabels );
+
+                for ( double nextMember : padded.getMembers() )
                 {
                     builder.addRight( nextMember );
                 }
@@ -146,35 +163,23 @@ public class SpaghettiPlot implements Metric<Pool<TimeSeries<Pair<Double, Ensemb
     }
 
     /**
-     * Uncovers the ensemble names from the series or generates some default names.
+     * Validates the presence of labels across all ensembles.
      *
-     * @param series the series
-     * @return the labels
+     * @param timeSeries the time-series
      */
 
-    private List<String> getEnsembleLabels( List<TimeSeries<Pair<Double, Ensemble>>> series )
+    private void validateLabels( List<TimeSeries<Pair<Double, Ensemble>>> timeSeries )
     {
-        Ensemble ensemble = series.get( 0 )
-                                  .getEvents()
-                                  .first()
-                                  .getValue()
-                                  .getRight();
-        List<String> returnMe;
-        if ( ensemble.hasLabels() )
+        if ( timeSeries.stream()
+                       .flatMap( t -> t.getEvents()
+                                       .stream() )
+                       .map( Event::getValue )
+                       .anyMatch( e -> !e.getRight()
+                                         .hasLabels() ) )
         {
-            returnMe = Arrays.asList( ensemble.getLabels().
-                                              getLabels() );
+            throw new MetricCalculationException( "Cannot calculate a spaghetti plot for ensemble time-series that do "
+                                                  + "not contain labels." );
         }
-        else
-        {
-            returnMe = new ArrayList<>();
-            for ( int i = 0; i < ensemble.size(); i++ )
-            {
-                returnMe.add( MEMBER + ( i + 1 ) );
-            }
-        }
-
-        return Collections.unmodifiableList( returnMe );
     }
 
     /**

--- a/wres-metrics/test/wres/metrics/MetricTestDataFactory.java
+++ b/wres-metrics/test/wres/metrics/MetricTestDataFactory.java
@@ -1247,11 +1247,14 @@ public final class MetricTestDataFactory
     {
         SortedSet<Event<Pair<Double, Ensemble>>> values = new TreeSet<>();
         values.add( Event.of( Instant.parse( "1985-03-13T00:00:00Z" ),
-                              Pair.of( 22.9, Ensemble.of( 22.8, 23.9 ) ) ) );
+                              Pair.of( 22.9, Ensemble.of( new double[] { 22.8, 23.9 },
+                                                          Ensemble.Labels.of( "1", "2" ) ) ) ) );
         values.add( Event.of( Instant.parse( "1985-03-13T06:00:00Z" ),
-                              Pair.of( 26.4, Ensemble.of( 23.8, 23.7 ) ) ) );
+                              Pair.of( 26.4, Ensemble.of( new double[] { 23.8, 23.7 },
+                                                          Ensemble.Labels.of( "1", "2" ) ) ) ) );
         values.add( Event.of( Instant.parse( "1985-03-13T12:00:00Z" ),
-                              Pair.of( 23.2, Ensemble.of( 16.1, 18.4 ) ) ) );
+                              Pair.of( 23.2, Ensemble.of( new double[] { 16.1, 18.4 },
+                                                          Ensemble.Labels.of( "1", "2" ) ) ) ) );
 
         TimeWindow inner = MessageUtilities.getTimeWindow( Instant.parse( FIRST_TIME ),
                                                            Instant.parse( SECOND_TIME ),
@@ -1278,6 +1281,64 @@ public final class MetricTestDataFactory
         Instant refTime = Instant.parse( "1985-03-12T18:00:00Z" );
         TimeSeriesMetadata metadata = MetricTestDataFactory.getBoilerplateMetadataWithT0( refTime );
         return builder.addData( TimeSeries.of( metadata, values ) )
+                      .setMetadata( meta )
+                      .build();
+    }
+
+    /**
+     * Returns a set of ensemble pairs with two time-series, each containing a different number of ensemble members and
+     * no baseline.
+     *
+     * @return ensemble pairs
+     */
+
+    public static wres.datamodel.pools.Pool<TimeSeries<Pair<Double, Ensemble>>> getTimeSeriesOfEnsemblePairsThree()
+    {
+        SortedSet<Event<Pair<Double, Ensemble>>> valuesOne = new TreeSet<>();
+        valuesOne.add( Event.of( Instant.parse( "1985-03-13T00:00:00Z" ),
+                                 Pair.of( 22.9, Ensemble.of( new double[] { 22.8, 23.9 },
+                                                             Ensemble.Labels.of( "1", "2" ) ) ) ) );
+        valuesOne.add( Event.of( Instant.parse( "1985-03-13T12:00:00Z" ),
+                                 Pair.of( 23.2, Ensemble.of( new double[] { 16.1, 18.4 },
+                                                             Ensemble.Labels.of( "1", "2" ) ) ) ) );
+
+        SortedSet<Event<Pair<Double, Ensemble>>> valuesTwo = new TreeSet<>();
+        valuesTwo.add( Event.of( Instant.parse( "1985-03-13T06:00:00Z" ),
+                                 Pair.of( 26.4, Ensemble.of( new double[] { 23.8, 23.7, 32.5 },
+                                                             Ensemble.Labels.of( "1", "2", "3" ) ) ) ) );
+
+        TimeWindow inner = MessageUtilities.getTimeWindow( Instant.parse( FIRST_TIME ),
+                                                           Instant.parse( SECOND_TIME ),
+                                                           Duration.ofHours( 24 ) );
+        TimeWindowOuter window = TimeWindowOuter.of( inner );
+
+        FeatureGroup featureGroup = Boilerplate.getFeatureGroup();
+
+        Evaluation evaluation = Evaluation.newBuilder()
+                                          .setRightVariableName( "MAP" )
+                                          .setMeasurementUnit( MM_DAY )
+                                          .build();
+
+        Pool pool = MessageFactory.getPool( featureGroup,
+                                            window,
+                                            null,
+                                            null,
+                                            false );
+
+        PoolMetadata meta = PoolMetadata.of( evaluation, pool );
+
+        Builder<TimeSeries<Pair<Double, Ensemble>>> builder = new Builder<>();
+
+        Instant refTime = Instant.parse( "1985-03-12T18:00:00Z" );
+        TimeSeriesMetadata metadata = MetricTestDataFactory.getBoilerplateMetadataWithT0( refTime );
+        TimeSeries<Pair<Double, Ensemble>> seriesOne = TimeSeries.of( metadata, valuesOne );
+
+        Instant refTimeTwo = Instant.parse( "1985-03-13T18:00:00Z" );
+        TimeSeriesMetadata metadataTwo = MetricTestDataFactory.getBoilerplateMetadataWithT0( refTimeTwo );
+        TimeSeries<Pair<Double, Ensemble>> seriesTwo = TimeSeries.of( metadataTwo, valuesTwo );
+
+        return builder.addData( seriesOne )
+                      .addData( seriesTwo )
                       .setMetadata( meta )
                       .build();
     }

--- a/wres-metrics/test/wres/metrics/timeseries/SpaghettiPlotTest.java
+++ b/wres-metrics/test/wres/metrics/timeseries/SpaghettiPlotTest.java
@@ -2,6 +2,7 @@ package wres.metrics.timeseries;
 
 import java.time.Duration;
 import java.time.Instant;
+import java.util.List;
 
 import com.google.protobuf.Timestamp;
 import org.apache.commons.lang3.tuple.Pair;
@@ -112,9 +113,107 @@ class SpaghettiPlotTest
         Pairs pairs = Pairs.newBuilder()
                            .addLeftVariableNames( DatasetOrientation.LEFT.toString()
                                                                          .toUpperCase() )
-                           .addRightVariableNames( "MEMBER 1" )
-                           .addRightVariableNames( "MEMBER 2" )
+                           .addRightVariableNames( "1" )
+                           .addRightVariableNames( "2" )
                            .addTimeSeries( timeSeries )
+                           .build();
+
+        PairsMetric metric = PairsMetric.newBuilder()
+                                        .setName( MetricName.SPAGHETTI_PLOT )
+                                        .setUnits( "MM/DAY" )
+                                        .build();
+
+        PairsStatistic expectedPairsStatistic = PairsStatistic.newBuilder()
+                                                              .setStatistics( pairs )
+                                                              .setMetric( metric )
+                                                              .build();
+
+        PairsStatisticOuter expected = PairsStatisticOuter.of( expectedPairsStatistic, expectedMetadata );
+
+        assertEquals( expected, actual );
+    }
+
+    @Test
+    void testApplyWithUnevenMembers()
+    {
+        // See GitHub #711
+
+        // Generate some data
+        Pool<TimeSeries<Pair<Double, Ensemble>>> input = MetricTestDataFactory.getTimeSeriesOfEnsemblePairsThree();
+
+        PairsStatisticOuter actual = this.plot.apply( input );
+
+        TimeWindow inner = MessageUtilities.getTimeWindow( Instant.parse( "1985-01-01T00:00:00Z" ),
+                                                           Instant.parse( "2010-12-31T11:59:59Z" ),
+                                                           Duration.ofHours( 24 ) );
+        TimeWindowOuter window = TimeWindowOuter.of( inner );
+
+        FeatureGroup featureGroup = Boilerplate.getFeatureGroup();
+
+        Evaluation evaluation = Evaluation.newBuilder()
+                                          .setRightVariableName( "MAP" )
+                                          .setMeasurementUnit( "MM/DAY" )
+                                          .build();
+
+        wres.statistics.generated.Pool pool = MessageFactory.getPool( featureGroup,
+                                                                      window,
+                                                                      null,
+                                                                      null,
+                                                                      false );
+
+        PoolMetadata expectedMetadata = PoolMetadata.of( evaluation, pool );
+
+
+        Timestamp firstTime = MessageUtilities.getTimestamp( Instant.parse( "1985-03-13T00:00:00Z" ) );
+        Timestamp secondTime = MessageUtilities.getTimestamp( Instant.parse( "1985-03-13T06:00:00Z" ) );
+        Timestamp thirdTime = MessageUtilities.getTimestamp( Instant.parse( "1985-03-13T12:00:00Z" ) );
+
+        Timestamp firstReferenceTime = MessageUtilities.getTimestamp( Instant.parse( "1985-03-12T18:00:00Z" ) );
+        ReferenceTime one = ReferenceTime.newBuilder()
+                                         .setReferenceTime( firstReferenceTime )
+                                         .setReferenceTimeType( ReferenceTime.ReferenceTimeType.T0 )
+                                         .build();
+
+        Timestamp secondReferenceTime = MessageUtilities.getTimestamp( Instant.parse( "1985-03-13T18:00:00Z" ) );
+        ReferenceTime two = ReferenceTime.newBuilder()
+                                         .setReferenceTime( secondReferenceTime )
+                                         .setReferenceTimeType( ReferenceTime.ReferenceTimeType.T0 )
+                                         .build();
+
+        Pairs.TimeSeriesOfPairs timeSeries =
+                Pairs.TimeSeriesOfPairs.newBuilder()
+                                       .addPairs( Pairs.Pair.newBuilder()
+                                                            .addLeft( 22.9 )
+                                                            .addRight( 22.8 )
+                                                            .addRight( 23.9 )
+                                                            .addRight( Double.NaN )
+                                                            .setValidTime( firstTime ) )
+                                       .addPairs( Pairs.Pair.newBuilder()
+                                                            .addLeft( 23.2 )
+                                                            .addRight( 16.1 )
+                                                            .addRight( 18.4 )
+                                                            .addRight( Double.NaN )
+                                                            .setValidTime( thirdTime ) )
+                                       .addReferenceTimes( one )
+                                       .build();
+
+        Pairs.TimeSeriesOfPairs timeSeriesTwo =
+                Pairs.TimeSeriesOfPairs.newBuilder()
+                                       .addPairs( Pairs.Pair.newBuilder()
+                                                            .addLeft( 26.4 )
+                                                            .addRight( 23.8 )
+                                                            .addRight( 23.7 )
+                                                            .addRight( 32.5 )
+                                                            .setValidTime( secondTime ) )
+                                       .addReferenceTimes( two )
+                                       .build();
+
+        Pairs pairs = Pairs.newBuilder()
+                           .addLeftVariableNames( DatasetOrientation.LEFT.toString()
+                                                                         .toUpperCase() )
+                           .addAllRightVariableNames( List.of( "1", "2", "3" ) )
+                           .addTimeSeries( timeSeries )
+                           .addTimeSeries( timeSeriesTwo )
                            .build();
 
         PairsMetric metric = PairsMetric.newBuilder()

--- a/wres-statistics/nonsrc/wresproto/pairs.proto
+++ b/wres-statistics/nonsrc/wresproto/pairs.proto
@@ -21,10 +21,16 @@ message Pairs
     // Zero or more time-series of pairs.
 
     repeated string left_variable_names = 2;
-    // The variable names of the left values, in order of appearance.
+    // The common variable names of the left values, in order of appearance.
+    // If the left variable names differ across time-series, then there are 
+    // no common labels and the left variable names are instead found in the 
+    // individual time-series.
 
     repeated string right_variable_names = 3;
-    // The variable names of the right values, in order of appearance.
+    // The variable names of the right values, in order of appearance. If the 
+    // right variable names differ across time-series, then there are no 
+    // common labels and the right variable names are instead found in the 
+    // individual time-series.
 
     message TimeSeriesOfPairs
     {
@@ -35,6 +41,14 @@ message Pairs
         
         repeated Pair pairs = 2;
         // One or more pairs.
+
+        repeated string left_variable_names = 3;
+        // The left variable names. Only present when they vary across 
+        // time-series.
+
+        repeated string right_variable_names = 4;
+        // The right variable names. Only present when they vary across 
+        // time-series.
     }
 
     message Pair 

--- a/wres-vis/src/wres/vis/data/ChartDataFactory.java
+++ b/wres-vis/src/wres/vis/data/ChartDataFactory.java
@@ -673,14 +673,6 @@ public class ChartDataFactory
     {
         Objects.requireNonNull( statistics );
 
-        List<String> leftNames = statistics.getStatistic()
-                                           .getStatistics()
-                                           .getLeftVariableNamesList();
-
-        List<String> rightNames = statistics.getStatistic()
-                                            .getStatistics()
-                                            .getRightVariableNamesList();
-
         int seriesNumber = 1;
         List<TimeSeries> allSeries = new ArrayList<>();
         for ( Pairs.TimeSeriesOfPairs nextSeries : statistics.getStatistic()
@@ -690,6 +682,16 @@ public class ChartDataFactory
             // Add a placeholder series for each variable
             List<TimeSeries> leftSeries = new ArrayList<>();
             List<TimeSeries> rightSeries = new ArrayList<>();
+
+            // Names aer either common to all series or vary by series, determine which. See GitHub #711
+            List<String> leftNames = ChartDataFactory.getVariableNames( statistics.getStatistic()
+                                                                                  .getStatistics()
+                                                                                  .getLeftVariableNamesList(),
+                                                                        nextSeries.getLeftVariableNamesList() );
+            List<String> rightNames = ChartDataFactory.getVariableNames( statistics.getStatistic()
+                                                                                   .getStatistics()
+                                                                                   .getRightVariableNamesList(),
+                                                                         nextSeries.getRightVariableNamesList() );
 
             // Create a placeholder time-series for each variable
             for ( String nextLeftName : leftNames )
@@ -1639,6 +1641,23 @@ public class ChartDataFactory
         Number leadDuration = DataUtilities.durationToNumericUnits( duration,
                                                                     durationUnits );
         return leadDuration.toString();
+    }
+
+    /**
+     * Returns whichever variable names exist.
+     * @param commonNames the common names
+     * @param seriesNames the series names
+     * @return the names that exist
+     */
+
+    private static List<String> getVariableNames( List<String> commonNames, List<String> seriesNames )
+    {
+        if ( !commonNames.isEmpty() )
+        {
+            return commonNames;
+        }
+
+        return seriesNames;
     }
 
     /**

